providers/base/tests/test_pipewire_utils.py

```diff
@@ -0,0 +1,634 @@
+
+        # parameters error
+        self.assertEqual(PipewireTestError.NO_AVAILABLE_PORT,
+                         pt.select_device("xxx", "sink", "hdmi"))
+        self.assertEqual(PipewireTestError.NO_AVAILABLE_PORT,
+                         pt.select_device("video", "xxx", "hdmi"))
+
+        mock_checkout.return_value = self.sink_audio_node
+        self.assertEqual(PipewireTestError.NO_AVAILABLE_PORT,
+                         pt.select_device("video", "sink", "hdmi"))
+
+    @patch("builtins.input")
+    @patch("subprocess.check_output")
+    def test_succ(self, mock_checkout, mock_input):
+        mock_checkout.return_value = self.sink_audio_node
+        mock_input.side_effect = ["t", 29]
+        pt = PipewireTest()
+        self.assertEqual(PipewireTestError.NO_ERROR,
+                         pt.select_device("audio", "sink", "hdmi"))
+
+    @patch("builtins.input")
+    @patch("subprocess.check_output")
+    def test_exit(self, mock_checkout, mock_input):
+        mock_checkout.return_value = self.sink_audio_node
+        mock_input.side_effect = ["t", "-1"]
+        pt = PipewireTest()
+        self.assertEqual(PipewireTestError.NO_ERROR,
+                         pt.select_device("audio", "sink", "hdmi"))
+
+
+class GstPipeLineTests(unittest.TestCase):
+
+    # Correct Device
+    device = '[{\
+                 "id": 29,\
+                 "type": "PipeWire:Interface:Device",\
+                 "version": 3,\
+                 "permissions": [ "r", "w", "x", "m" ],\
+                 "info": {\
+                   "max-input-ports": 0,\
+                   "max-output-ports": 0,\
+                   "change-mask": [ "input-ports",\
+                                    "output-ports",\
+                                    "state", "props",\
+                                    "params" ],\
+                   "n-input-ports": 0,\
+                   "n-output-ports": 0,\
+                   "state": "suspended",\
+                   "error": null,\
+                   "props": {\
+                     "media.class": "Audio/Device",\
+                     "factory.name": "support.node.driver",\
+                     "node.name": "Freewheel-Driver",\
+                     "node.description": "unit test",\
+                     "priority.driver": 19000,\
+                     "node.group": "pipewire.freewheel",\
+                     "node.freewheel": true,\
+                     "factory.id": 10,\
+                     "clock.quantum-limit": 8192,\
+                     "node.driver": true,\
+                     "object.id": 29,\
+                     "object.serial": 29\
+                   },\
+                   "params": {\
+                       "Route": [{\
+                           "name": "hdmi_demo_output",\
+                           "available": "yes",\
+                           "description": "hdmi demo output"\
+                       }]\
+                   }\
+                 }\
+                }]'
```

Comment:
please use multiline strings

Reply:
Change this to suggestion according to code review criteria, feel free to close it.